### PR TITLE
support freetube playlists import/export

### PR DESCRIPTION
**Pull request by yoguut**
Closes https://github.com/libre-tube/LibreTube/issues/3816

[x] tested import FreeTube playlists to app
[x] tested export Piped / LibreTube playlists from app
[x] tested export FreeTube playlists from app

### CODE REVIEW

File: app/src/main/java/com/github/libretube/api/PlaylistsHelper.kt [hunk 1]
```diff
@@ -167,18 +203,33 @@ object PlaylistsHelper {
...
     }
 
-    suspend fun exportPlaylists(): List<ImportPlaylist> = withContext(Dispatchers.IO) {
+    suspend fun exportPipedPlaylists(): List<PipedImportPlaylist> = withContext(Dispatchers.IO) {
         getPlaylists()
             .map { async { getPlaylist(it.id!!) } }
             .awaitAll()
             .map {
                 val videos = it.relatedStreams.map { item ->
                     "$YOUTUBE_FRONTEND_URL/watch?v=${item.url!!.toID()}"
```
Comment (yoguut) [inline app/src/main/java/com/github/libretube/api/PlaylistsHelper.kt:212]: Question to reviewer: why force unwrap optional `url`?

**Comment by Bnyro** [comment]
Maybe we can rename the string of the Piped option to "Piped / LibreTube"? Otherwise users that exported it earlier probably don't know which format they exported it to.

File: app/src/main/java/com/github/libretube/api/PlaylistsHelper.kt (hunk 1 above)
Comment (Bnyro) [reply app/src/main/java/com/github/libretube/api/PlaylistsHelper.kt:212]: `url` is nullable and `toID()` only works with non-null strings. So we either need to use `!!` or `?`. Which one we use doesn't matter, because it should never become null. But asserting it as non-null is better in this case because `?` indicates it can be null.

File: app/src/main/java/com/github/libretube/api/PlaylistsHelper.kt [hunk 2]
```diff
@@ -133,7 +135,41 @@ object PlaylistsHelper {
         }
     }
 
-    suspend fun importPlaylists(playlists: List<ImportPlaylist>) = withContext(Dispatchers.IO) {
+    suspend fun importFreeTubePlaylists(playlists: List<FreeTubeImportPlaylist>) = withContext(Dispatchers.IO) {
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/api/PlaylistsHelper.kt:138]: That's just a lot of duplicated code. Instead, please keep the old `importPlaylists` function and call it no matter if it's a Piped or a FreeTube playlist that's being imported.
That means we should "convert" the `List<FreeTubeImportPlaylist` to a `List<PipedImportPlaylist>` using 
```kotlin
freeTubePlaylists.map { PipedImportPlaylist(it.title, ...) }
```
(the above is just a simple example).

File: app/src/main/java/com/github/libretube/helpers/ImportHelper.kt [hunk 3]
```diff
@@ -144,47 +189,49 @@ object ImportHelper {
...
+                }
+                try {
+                    PlaylistsHelper.importNewPipePlaylists(importPlaylists)
+                    activity.toastFromMainDispatcher(R.string.success)
+                } catch (e: Exception) {
+                    Log.e(TAG(), e.toString())
+                    e.localizedMessage?.let {
+                        activity.toastFromMainDispatcher(it)
+                    }
+                }
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/helpers/ImportHelper.kt:205]: That try catch block is being called three times here (for each playlist type). Instead, it should only be there once after the `when` clause (which will work after applying my suggestion from `PlaylistsHelper.kt`.

File: app/src/main/java/com/github/libretube/helpers/ImportHelper.kt [hunk 4]
```diff
@@ -76,6 +77,8 @@ object ImportHelper {
                     }
                 }.orEmpty()
             }
+            // Piped case not use for subscriptions
+            ImportFormat.PIPED -> emptyList()
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/helpers/ImportHelper.kt:81]:
```suggestion
            else -> throw IllegalArgumentException()
```

File: app/src/main/java/com/github/libretube/helpers/ImportHelper.kt [hunk 5]
```diff
@@ -123,12 +126,38 @@ object ImportHelper {
...
+                importPlaylists.forEach { playlist ->
+                    playlist.videos = playlist.videos.map { it.takeLast(11) }
+                }
+            }
+            ImportFormat.FREETUBE -> {
+                val playlistFile = activity.contentResolver.openInputStream(uri)?.use {
+                    JsonHelper.json.decodeFromStream<List<FreeTubeImportPlaylist>>(it)
+                }
+                val playlists = playlistFile?.map { playlist ->
+                    // convert FreeTube videos to listOf string
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/helpers/ImportHelper.kt:149]:
```suggestion
                    // convert FreeTube videos to list of string
```

File: app/src/main/java/com/github/libretube/helpers/ImportHelper.kt [hunk 6]
```diff
@@ -144,23 +173,13 @@ object ImportHelper {
...
-                val message = activity.getString(R.string.unsupported_file_format, fileType)
-                activity.toastFromMainDispatcher(message)
-                return
             }
-        }
-
-        // convert the YouTube URLs to videoIds
-        importPlaylists.forEach { playlist ->
-            playlist.videos = playlist.videos.map { it.takeLast(11) }
+            ImportFormat.NEWPIPE -> Unit
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/helpers/ImportHelper.kt:182]:
```suggestion
            else -> throw IllegalArguemntException()
```

File: app/src/main/java/com/github/libretube/api/PlaylistsHelper.kt [hunk 7]
```diff
@@ -133,52 +139,69 @@ object PlaylistsHelper {
...
-                    playlist.videos.mapIndexed { index, id -> id to index }
-                        .groupBy { it.second % 20 }.forEach { (_, videos) ->
-                            videos.map {
-                                async {
-                                    runCatching {
-                                        val stream = RetrofitInstance.api.getStreams(it.first).toStreamItem(
-                                            it.first,
-                                        )
-                                        addToPlaylist(playlistId, stream)
+    suspend fun importPlaylists(playlists: List<PipedImportPlaylist>) =
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/api/PlaylistsHelper.kt:142]: Please try to undo these formatting changes since they are not related to your PR really

**Comment by Bnyro** [comment]
Only some minor things from my side, looks good to me otherwise!

**Review by Bnyro — APPROVED** [review]

**Comment by Bnyro** [comment]
Have you tested whether it works well after the latest changes?

**Comment by yoguut** [comment]
> Have you tested whether it works well after the latest changes?

_Yes and kind of. I don't have Piped account or Youtube account. So my playlists come from FreeTube export. And then once imported the FreeTube playlists in the app, I exported it to Piped & FreeTube formats._

_It would be great if someone has either an Piped account or Youtube account to help testing by exporting playlists, then importing into the app to ensure the most current formats of these 2 services are consistent to FreeTube's export formats._


[x] tested import FreeTube playlists to app
[x] tested export Piped / LibreTube playlists from app
[x] tested export FreeTube playlists from app

**Comment by Bnyro** [comment]
Tested the Piped format and the YouTube CSV and works as expected. Thanks for your PR!
